Report bad tightening/loosening of attributes in OHI (#41336)
dotnet/roslyn: docs/features/nullable-reference-types.md

When simple pre- and post-condition attributes are applied to a property, and an allowed input state is assigned to the property, the property's state is updated to be an allowed output state. For instance an `[AllowNull] string` property can be assigned `null` and still return not-null values.

Enforcing nullability attributes within method bodies:

> - An input parameter marked with [AllowNull] is initialized with a maybe-null (or maybe-default) state.

Reviewer: ```suggestion
- An input parameter marked with `[AllowNull]` is initialized with a maybe-null (or maybe-default) state.
```

> - An input parameter marked with [DisallowNull] is initialized with a not-null state.

Reviewer: ```suggestion
- An input parameter marked with `[DisallowNull]` is initialized with a not-null state.
```

> - A parameter marked with [MaybeNull] or [MaybeNullWhen] can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with [NotNullWhen] (the attribute is ignored).

Reviewer: ```suggestion
- A parameter marked with `[MaybeNull]` or `[MaybeNullWhen]` can be assigned a maybe-null value, without warning. Same for return values. Same for a nullable parameter marked with `[NotNullWhen]` (the attribute is ignored).
```

> - A parameter marked with [NotNull] will produce a warning when assigned a maybe-null value. Same for return values.

Reviewer: This only applies to `out` and `ref` parameters right?

```suggestion
- A by-reference parameter marked with `[NotNull]` will produce a warning when assigned a maybe-null value. Same for return values.
```

Author: @sharwell, no this applies to by-value parameters as well. I'll update test `UnconditionalAttributesAffectMethodBodies_ReferenceType` to show that behavior.

---
In reply to: [373548582](https://github.com/dotnet/roslyn/pull/41336#discussion_r373548582) [](ancestors = 373548582)

> - The state of a parameter marked with [Maybe/NotNullWhen] is checked upon exiting the method instead.

Reviewer: ```suggestion
- The state of a parameter marked with `[MaybeNullWhen]`/`[NotNullWhen]` is checked upon exiting the method instead.
```

> - The state of a parameter marked with `[MaybeNullWhen]`\`[NotNullWhen]` is checked upon exiting the method instead.

Reviewer 2: >\ [](start = 56, length = 1)

"/" or "or"

> - A method marked with [DoesNotReturn] will produce a warning if it returns or exits normally.

Reviewer: ```suggestion
- A method marked with `[DoesNotReturn]` will produce a warning if it returns or exits normally.
```

> Note: we don't validate the internal consistency of auto-properties, so it is possible to misused attributes on auto-props as on fields. For example: `[AllowNull, NotNull] public TOpen P { get; set; }`.

Reviewer: ```suggestion
Note: we don't validate the internal consistency of auto-properties, so it is possible to misuse attributes on auto-props as on fields. For example: `[AllowNull, NotNull] public TOpen P { get; set; }`.
```

> Note: we don't validate the internal consistency of auto-properties, so it is possible to misused attributes on auto-props as on fields. For example: `[AllowNull, NotNull] public TOpen P { get; set; }`.

Reviewer: 💡 Consider breaking out a section specifically to point out things that are _not_ checked

Enforcing nullability attributes when overriding and implementing:
In addition to checking the types in overrides/implementations are compatible with overridden/implemented members, we also check that the nullability attributes are compatible.

> - For input parameters (by-value and `in`), we check that the worst allowed value by the overridden/implemented parameter can be assigned to the overriding/implementing parameter.

Reviewer: ```suggestion
- For input parameters (by-value and `in`), we check that the widest allowed value by the overridden/implemented parameter can be assigned to the overriding/implementing parameter.
```

> - For output parameters (`out` and return values), we check that the worst produced value by the overriding/implementing parameter can be assigned to the overridden/implemented parameter.

Reviewer: ```suggestion
- For output parameters (`out` and return values), we check that the widest produced value by the overriding/implementing parameter can be assigned to the overridden/implemented parameter.
```

- For `ref` parameters and return values, we check both.

> - We check that the post-condition contract `[Not/MaybeNull]` on input parameters is enforced by overriding/implementing members.

Reviewer: ```suggestion
- We check that the post-condition contract `[NotNull]`/`[MaybeNull]` on input parameters is present and enforced by overriding/implementing members.
```

> - We check that overrides/implementations have the `[DoesNotReturn]` attribute if the overridden/implemented member has it.

Reviewer: ❔ Do we allow an implementation to use `[DoesNotReturn]` if the original definition used `[DoesNotReturnIf]`?

Author: I'll add a test, but I think we allow it and that behavior is correct. 
The implementation abides by the overridden/implemented method's contract.

---
In reply to: [373550920](https://github.com/dotnet/roslyn/pull/41336#discussion_r373550920) [](ancestors = 373550920)

Author: It turns out the problem doesn't even present itself. `[DoesNotReturn]` is applied to methods, but `[DoesNotReturnIf(...)]` is applied to parameters. They don't cross paths.
